### ISSUE TITLE
How does the release pipeline work?

### ISSUE THREAD

**Issue by ba32107** [summerwind/actions-runner-controller#279]
Hi, I am currently working on trying this controller in our in-house Kubernetes cluster. We are working in a locked-down environment so I won't be able to deploy this component using the official releases - I need to tweak and modify it in several places to make sure it will work in our cluster.

I'd like to understand how exactly the pipeline works. I am specifically interested in the Kubernetes resources, so for now I'll ignore the docker images. I have been trying to figure it out, but it's still not completely clear to me. My current understanding is the following:
 * By creating a GitHub release, the action `release.yml` is triggered
 * This action invokes the Makefile via `make github-release`
 * In the Makefile, `manifests-118` generates the CRD definitions from the Go source code via `controller-gen` (at least I _think_ that's what happens - not familiar with `controller-gen`)
 * These files are then processed, then copied under the `charts` directory
 * The final Kubernetes spec is generated via Kustomize, in the `config/default` directory, then uploaded to the GitHub release.
 * In parallel to all the above, whenever files are changed under the `charts` directory, the action `on-push-master-publish-chart.yml` will trigger
 * This action uses `chart-releaser` tool to generate Helm charts. I am not familiar with this tool, but I presume it reads the `charts` directory and uploads stuff to the GitHub release. 

Questions:
 * Even though files are being generated, I don't see them being committed back to GitHub anywhere, yet the files are in the repo. Was this done manually? 
 * What is the difference between the two artifacts generated? There is `actions-runner-controller.yaml` generated by Kustomize (in the Makefile), and there are also Helm charts generated by a completely different GitHub Action. What is the difference between these?
 * If I'm not mistaken, the answer to this question is that they are simply two distribution methods for the same application: I can use either Helm to deploy this app or simple `kubectl` commands. However, there seem to be a different set of files in the `charts` and `config` directories. I don't see how they produce the same final Kubernetes spec. 

To summarize, I'd like to generate the Kubernetes manifests myself, and not use the official releases. What is the easiest way to do this?

**Comment by callum-tait-pbx** [comment]
> Even though files are being generated, I don't see them being committed back to GitHub anywhere, yet the files are in the repo. Was this done manually?

The Makefile is mostly testing or tear up / down of an env for acceptance testing

> What is the difference between the two artifacts generated? There is actions-runner-controller.yaml generated by Kustomize (in the Makefile), and there are also Helm charts generated by a completely different GitHub Action. What is the difference between these?

The Helm release is quite new. The repository originally started life as only a Kutomize build, the community wanted a Helm chart as this is the much more popular tool. The `chart.yaml` workflow is very new. In theory there isn't really any differences between the default Helm chart and the default Kustomize configuration and I think this is still true but I'm not 100% on that as I work in a Helm shop so only really look at that part of the repo.

> If I'm not mistaken, the answer to this question is that they are simply two distribution methods for the same application: I can use either Helm to deploy this app or simple kubectl commands. However, there seem to be a different set of files in the charts and config directories. I don't see how they produce the same final Kubernetes spec.

All the k8s bit is under the `config` folder for Kustomize and `charts/actions-runner-controller` for Helm. There should really be only place for this but becuase of how the solution started off as Kustomize based and Helm was added much latter there are 2 sources of truth. This for obvious reasons is not ideal however most of the changes that happen are in the Golang app rather than at the Kubernetes level so its not as terrible as it first seems.

If you are using Helm then there aren't really any manifests to generate as the chart is the manifest. You simply customise it for you env via the `values.yaml` file. If you need to extend the chart then you would simply make changes to the manifests in the templates folder.
Easiest way to actually generate a manifest for yourself would be to build to run the build through Kustomize. This is very easy to do, install kustomize however you want to do that e.g. via brew.

```shell
cd config
kustomize build ./default > manifest.yaml
```

Kustomize will stitch together the patches etc into a single manifest.yaml file which you can deploy.

**Comment by ba32107** [comment]
Hi @callum-tait-pbx, excellent, thanks, your reply just confirmed my understanding. All good!
